src/masternodes/mn_checks.cpp

```diff
@@ -1029,6 +1114,443 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
+        if (order.orderType != CICXOrder::TYPE_INTERNAL && order.orderType != CICXOrder::TYPE_EXTERNAL)
+            return Res::Err("invalid order type!");
+
+        if (order.orderType == CICXOrder::TYPE_INTERNAL)
+        {
+            auto tokenFrom = mnview.GetToken(order.idToken);
+            if (!tokenFrom)
+                return Res::Err("%s: %s", __func__, "tokenFrom (" +  tokenFrom->CreateSymbolKey(order.idToken) + ") does not exist!");
+
+            if (order.ownerAddress.empty())
+                return Res::Err("ownerAddress must be not-null!");
+
+            // subtract the balance from tokenFrom to dedicate them for the order
+            CScript txidAddr(order.creationTx.begin(), order.creationTx.end());
+            res = ModifyTokenBalance(order.idToken, order.amountFrom, order.ownerAddress, txidAddr);
+            if (!res.ok)
+                return Res::Err("%s: %s", __func__, res.msg);
+        }
+        else if (order.orderType == CICXOrder::TYPE_EXTERNAL)
+        {
+            auto tokenTo = mnview.GetToken(order.idToken);
+            if (!tokenTo)
+                return Res::Err("%s: %s", __func__, "tokenTo id (" +  tokenTo->CreateSymbolKey(order.idToken) + ") does not exist!");
+        }
+
+        if (order.amountFrom == 0)
+            return Res::Err("order amountFrom must be greater than 0!");
+        if (order.amountToFill != order.amountFrom)
+            return Res::Err("order amountToFill does not equal to amountFrom!");
+        if (order.orderPrice == 0)
+            return Res::Err("order price must be greater than 0!");
+        if (order.expiry == 0)
+            return Res::Err("order expiry must be greater than 0!");
+
+        return mnview.ICXCreateOrder(order);
+    }
+
+    Res operator()(const CICXMakeOfferMessage& obj) const {
+        auto res = CheckICXTx();
+        if (!res)
+            return res;
+
+        CICXMakeOfferImplemetation makeoffer;
+        static_cast<CICXMakeOffer&>(makeoffer) = obj;
+
+        makeoffer.creationTx = tx.GetHash();
+        makeoffer.creationHeight = height;
+
+        auto order=mnview.GetICXOrderByCreationTx(makeoffer.orderTx);
+        if (!order)
+            return Res::Err("%s: %s", __func__, "order with creation tx " + makeoffer.orderTx.GetHex() + " does not exists!");
+        
+        CAmount calcedAmount(static_cast<CAmount>((arith_uint256(order->amountToFill) * arith_uint256(order->orderPrice) / arith_uint256(COIN)).GetLow64()));
+        if (calcedAmount < makeoffer.amount)
+            makeoffer.amount = calcedAmount;
+
+        CScript txidAddr(makeoffer.creationTx.begin(),makeoffer.creationTx.end());
+        
+        //calculating DFI per BTC
+        auto id = mnview.ICXGetDFIBTCPoolPairId(height);
+        auto DFIBTCPoolPair = mnview.GetPoolPair(id);
+        CAmount DFIperBTC = CICXOrderView::DEFAULT_DFI_BTC_PRICE;
+        if (DFIBTCPoolPair)
```

Comment:
If didn't get the poolpair, do we still need to continue? I think we should just return an error and stop here.

Reply:
We have a default price for `DFIperBTC` for this case. But there should be gov variable that will set pool pair id.